app/models/vm/operations.rb

```diff
@@ -14,7 +14,11 @@ module Vm::Operations
   end
 
   def cockpit_url
-    URI::HTTP.build(:host => ipaddresses.first, :port => 9090).to_s
+    URI::HTTP.build(:host => return_ipv4_address, :port => 9090).to_s
+  end
+
+  def return_ipv4_address
```

Comment:
Why such a strange name?  Why not the simpler `ipv4_address`?  Also, if `ipaddresses.first` is returned and it is not an ipv4 address -- I think that should be part of the caller and not part of this method.  So, `cockpit_url` method would have `:host => ipv4_address || ipaddresses.first`